File: backend/app/routers/anime.py
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional
import aiosqlite
from datetime import datetime, timezone
from pydantic import BaseModel

from ..database import get_db
from ..schemas import AnimeCreate, AnimeUpdate, AnimeResponse, AnimeStatus


class ExportData(BaseModel):
    """导出数据的格式"""
    version: int = 1
    exported_at: str = ""
    anime_list: list[AnimeCreate]


class ImportResult(BaseModel):
    """导入结果的格式"""
    imported: int = 0
    updated: int = 0
    skipped: int = 0
    errors: list[str] = []


class ImportRequest(BaseModel):
    """导入请求的格式"""
    anime_list: list[AnimeCreate]

router = APIRouter(prefix="/api/anime", tags=["anime"])


@router.get("", response_model=list[AnimeResponse])
async def list_anime(
    status: Optional[AnimeStatus] = None,
    search: Optional[str] = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: aiosqlite.Connection = Depends(get_db),
):
    query = "SELECT * FROM anime WHERE 1=1"
    params = []

    if status:
        query += " AND status = ?"
        params.append(status.value)
    if search:
        query += " AND title LIKE ?"
        params.append(f"%{search}%")

    query += " ORDER BY updated_at DESC LIMIT ? OFFSET ?"
    params.extend([page_size, (page - 1) * page_size])

    cursor = await db.execute(query, params)
    rows = await cursor.fetchall()
    return [AnimeResponse(**dict(row)) for row in rows]


@router.get("/{anime_id}", response_model=AnimeResponse)
async def get_anime(anime_id: int, db: aiosqlite.Connection = Depends(get_db)):
    cursor = await db.execute("SELECT * FROM anime WHERE id = ?", (anime_id,))
    row = await cursor.fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Anime not found")
    return AnimeResponse(**dict(row))


@router.post("", response_model=AnimeResponse, status_code=201)
async def create_anime(anime: AnimeCreate, db: aiosqlite.Connection = Depends(get_db)):
    # BUG-3 修复：检查 bangumi_id 是否已存在，防止重复添加
    if anime.bangumi_id is not None:
        cursor = await db.execute(
            "SELECT id FROM anime WHERE bangumi_id = ?", (anime.bangumi_id,)
        )
        existing = await cursor.fetchone()
        if existing:
            raise HTTPException(status_code=409, detail="该番剧已存在于你的列表中")

    cursor = await db.execute(
        """INSERT INTO anime (title, cover_url, description, total_episodes, current_episode, status, score, air_date, bangumi_id)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        (
            anime.title,
            anime.cover_url,
            anime.description,
            anime.total_episodes,
            anime.current_episode,
            anime.status.value,
            anime.score,
            anime.air_date,
            anime.bangumi_id,
        ),
    )
    await db.commit()
    new_id = cursor.lastrowid
    return await get_anime(new_id, db)


@router.put("/{anime_id}", response_model=AnimeResponse)
async def update_anime(anime_id: int, anime: AnimeUpdate, db: aiosqlite.Connection = Depends(get_db)):
    existing = await get_anime(anime_id, db)
    if not existing:
        raise HTTPException(status_code=404, detail="Anime not found")

    fields = anime.model_dump(exclude_unset=True)
    if not fields:
        return existing

    if "status" in fields and fields["status"]:
        fields["status"] = fields["status"].value

    set_clause = ", ".join(f"{k} = ?" for k in fields)
    set_clause += ", updated_at = datetime('now')"
    values = list(fields.values()) + [anime_id]

    await db.execute(f"UPDATE anime SET {set_clause} WHERE id = ?", values)
    await db.commit()
    return await get_anime(anime_id, db)


@router.delete("/{anime_id}", status_code=204)
async def delete_anime(anime_id: int, db: aiosqlite.Connection = Depends(get_db)):
    cursor = await db.execute("SELECT id FROM anime WHERE id = ?", (anime_id,))
    if not await cursor.fetchone():
        raise HTTPException(status_code=404, detail="Anime not found")
    await db.execute("DELETE FROM anime WHERE id = ?", (anime_id,))
    await db.commit()


@router.get("/export/all", response_model=ExportData)
async def export_anime(db: aiosqlite.Connection = Depends(get_db)):
    """导出所有番剧数据为 JSON"""
    cursor = await db.execute("SELECT * FROM anime ORDER BY updated_at DESC")
    rows = await cursor.fetchall()
    anime_list = [AnimeCreate(**dict(row)) for row in rows]
    return ExportData(
        version=1,
        exported_at=datetime.now(timezone.utc).isoformat(),
        anime_list=anime_list,
    )


@router.post("/import", response_model=ImportResult)
async def import_anime(
    payload: ImportRequest,
    db: aiosqlite.Connection = Depends(get_db),
):
    """导入番剧数据（按 bangumi_id 匹配，存在则更新，不存在则新增）"""
    result = ImportResult()

    for item in payload.anime_list:
        try:
            # 尝试按 bangumi_id 匹配已存在的记录
            if item.bangumi_id is not None:
                cursor = await db.execute(
                    "SELECT id FROM anime WHERE bangumi_id = ?", (item.bangumi_id,)
                )
                existing = await cursor.fetchone()

                if existing:
                    # 更新现有记录
                    await db.execute(
                        """UPDATE anime
                           SET title = ?, cover_url = ?, description = ?,
                               total_episodes = ?, current_episode = ?,
                               status = ?, score = ?, air_date = ?,
                               updated_at = datetime('now')
                           WHERE id = ?""",
                        (
                            item.title,
                            item.cover_url,
                            item.description,
                            item.total_episodes,
                            item.current_episode,
                            item.status.value,
                            item.score,
                            item.air_date,
                            existing["id"],
                        ),
                    )
                    result.updated += 1
                    continue

            # 不存在则新增
            await db.execute(
                """INSERT INTO anime
                   (title, cover_url, description, total_episodes, current_episode, status, score, air_date, bangumi_id)
                   VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)""",
                (
                    item.title,
                    item.cover_url,
                    item.description,
                    item.total_episodes,
                    item.current_episode,
                    item.status.value,
                    item.score,
                    item.air_date,
                    item.bangumi_id,
                ),
            )
            result.imported += 1

        except Exception as e:
            result.errors.append(f"「{item.title}」导入失败: {str(e)}")
            result.skipped += 1

    await db.commit()
    return result

File: anisync_flutter/lib/models/anime.dart
/// 番剧数据模型，与后端 AnimeResponse 完全对齐
class Anime {
  final int id;
  final String title;
  final String? coverUrl;
  final String? description;
  final int totalEpisodes;
  final int currentEpisode;
  final String status; // plan | watching | completed
  final int? score;
  final String? airDate;
  final int? bangumiId;
  final String? createdAt;
  final String? updatedAt;

  Anime({
    required this.id,
    required this.title,
    this.coverUrl,
    this.description,
    this.totalEpisodes = 0,
    this.currentEpisode = 0,
    this.status = 'plan',
    this.score,
    this.airDate,
    this.bangumiId,
    this.createdAt,
    this.updatedAt,
  });

  factory Anime.fromJson(Map<String, dynamic> json) {
    return Anime(
      id: json['id'] as int,
      title: json['title'] as String,
      coverUrl: json['cover_url'] as String?,
      description: json['description'] as String?,
      totalEpisodes: json['total_episodes'] as int? ?? 0,
      currentEpisode: json['current_episode'] as int? ?? 0,
      status: json['status'] as String? ?? 'plan',
      score: json['score'] as int?,
      airDate: json['air_date'] as String?,
      bangumiId: json['bangumi_id'] as int?,
      createdAt: json['created_at'] as String?,
      updatedAt: json['updated_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'cover_url': coverUrl,
      'description': description,
      'total_episodes': totalEpisodes,
      'current_episode': currentEpisode,
      'status': status,
      'score': score,
      'air_date': airDate,
      'bangumi_id': bangumiId,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }

  /// 用于导出全量数据
  Map<String, dynamic> toExportJson() {
    return {
      'title': title,
      'cover_url': coverUrl,
      'description': description,
      'total_episodes': totalEpisodes,
      'current_episode': currentEpisode,
      'status': status,
      'score': score,
      'air_date': airDate,
      'bangumi_id': bangumiId,
    };
  }

  /// 是否已看完
  bool get isCompleted => totalEpisodes > 0 && currentEpisode >= totalEpisodes;

  /// 进度百分比 0~100
  double get progressPercent {
    if (totalEpisodes <= 0) return 0;
    return (currentEpisode / totalEpisodes * 100).clamp(0, 100);
  }

  /// 状态中文标签
  String get statusLabel {
    switch (status) {
      case 'watching':
        return '在看';
      case 'completed':
        return '已看完';
      case 'plan':
      default:
        return '想看';
    }
  }

  Anime copyWith({
    int? id,
    String? title,
    String? coverUrl,
    String? description,
    int? totalEpisodes,
    int? currentEpisode,
    String? status,
    int? score,
    String? airDate,
    int? bangumiId,
    String? createdAt,
    String? updatedAt,
  }) {
    return Anime(
      id: id ?? this.id,
      title: title ?? this.title,
      coverUrl: coverUrl ?? this.coverUrl,
      description: description ?? this.description,
      totalEpisodes: totalEpisodes ?? this.totalEpisodes,
      currentEpisode: currentEpisode ?? this.currentEpisode,
      status: status ?? this.status,
      score: score ?? this.score,
      airDate: airDate ?? this.airDate,
      bangumiId: bangumiId ?? this.bangumiId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
